### PR TITLE
Phase 2 CryptoContext

### PR DESCRIPTION
**Pull request by reneroliveira**
#### Homomorphic Operations

- EvalNegate
- EvalNegateInPlace
- EvalAddInPlace
- EvalAddMultable
- EvalAddMultableInPlace
- EvalSubInPlace
- EvalSubMutable
- EvalSubMutableInplace
- EvalMultMutable
- EvalMultMutableInPlace
- EvalSquare
- EvalSquareMutable
- EvalSquareInPlace
- EvalMultnoRelin
- Relinearize
- RelinearizeInPlace
- EvalMultAndRelinearize

#### Automorphisms

- FindAutomorphismIndex
- FindAutomorphismIndices
- EvalFastRotationExt

#### Key Generation

- EvalMultKeysGen
- EvalAtIndexKeyGen
- EvalSumKeyGen
- EvalSumRowsKeyGen
- EvalSumColsKeyGen

#### Polynomial Operations

- EvalAtIndex
- EvalPolyLinear
- EvalPolyPS
- EvalChebyshevSeries
- EvalChebyshevSeriesLinear
- EvalChebyshevSeriesPS

#### Math/Trigonometric Functions

- EvalSin
- EvalCos
- EvalLogistic
- EvalDivide

#### Matrix Operations

- EvalSumRows
- EvalSumCols
- EvalInnerProduct

#### Plaintext Conversion

- MakeStringPlaintext
- MakeCoefPackedPlaintext

### CODE REVIEW

File: src/bindings.cpp [hunk 1]
```diff
@@ -100,65 +103,149 @@ void bind_crypto_context(py::module &m)
...
-            py::arg("depth") = static_cast<size_t>(1),
-            py::arg("level") = static_cast<uint32_t>(0),
-            py::arg("params") = py::none(),
-            py::arg("slots") = 0)
+        .def("MakeCoefPackedPlaintext", &CryptoContextImpl<DCRTPoly>::MakeCoefPackedPlaintext)
+        .def("MakeCKKSPackedPlaintext", static_cast<Plaintext (CryptoContextImpl<DCRTPoly>::*)(const std::vector<std::complex<double>> &, size_t, uint32_t, const std::shared_ptr<ParmType>, usint) const>(&CryptoContextImpl<DCRTPoly>::MakeCKKSPackedPlaintext), "Make a CKKS plaintext from a vector of complex doubles",
+             py::arg("value"),
+             py::arg("depth") = static_cast<size_t>(1),
+             py::arg("level") = static_cast<uint32_t>(0),
+             py::arg("params") = py::none(),
```
Comment (IanQS) [inline src/bindings.cpp:117]: does this mean that we do not accept any params, even those that might be valid in a pure C++ context? 

Is this something we will address in the future? If yes, make a TODO and if not just add a comment justifying the design choice. Likewise for other places

**Review by IanQS — CHANGES_REQUESTED** [review]
LGTM overall. I did have a quick comment, but other than that it's looking good. What's your plan for this branch? Are you hoping to merge it in soon or will you push more changes to it?

**Comment by reneroliveira** [comment]
> LGTM overall. I did have a quick comment, but other than that it's looking good. What's your plan for this branch? Are you hoping to merge it in soon or will you push more changes to it?

From the cryptocontext part, this branch covers everything for the first release, and the dev branch will be complete. I have to make more small changes like code cleaning, comments, etc, but I don't know if I do it on this branch, or create another one.

File: src/bindings.cpp (hunk 1 above)
Comment (reneroliveira) [reply src/bindings.cpp:117]: > does this mean that we do not accept any params, even those that might be valid in a pure C++ context?
> 
> Is this something we will address in the future? If yes, make a TODO and if not just add a comment justifying the design choice. Likewise for other places

On the C++ side, params have a default of nullptr. py::none() is the analogous version in Python. I'll check if the python user can pass something different than that. If not, I'll add the TODO mark to address it in the future.

**Pull request merged**